### PR TITLE
Fix visibility checking of mutually recursive exports

### LINKED ISSUE #14753: Private types export check gets thrown off by mutually recursive types that are exported separately
**TypeScript Version:**  2.2.1 
**Code**

```ts
interface A {
  b: B
}

interface B {
  a: A
}
export {A, B}
```

**Expected behavior:**
No errors, all interfaces are exported.

**Actual behavior:**
f.ts(7,6): error TS4033: Property 'b' of exported interface has or is using private name 'B'.

### LINKED ISSUE #17085: Missing exports in declarations, when destructing is used
**TypeScript Version:**  2.4.1

**Code**

`src/foo.ts`:

```ts
const foo = { bar: 'hello' };
export { foo };
```

`src/index.ts`:

```ts
import { foo } from './foo';
export { foo };

const { bar: baz } = foo;
export { baz };
```

`package.json`:

```json
{
  "name": "ts-declaration",
  "version": "1.0.0",
  "private": true,
  "main": "dist/index.js",
  "scripts": {
    "build": "tsc"
  },
  "devDependencies": {
    "typescript": "^2.4.1"
  }
}
```

`tsconfig.json`:

```json
{
  "compilerOptions": {
    "module": "es2015",
    "moduleResolution": "node",
    "noImplicitAny": true,
    "noImplicitReturns": true,
    "noImplicitThis": true,
    "strictNullChecks": true,
    "target": "es5",
    "outDir": "dist/",
    "declaration": true
  },
  "exclude": [
    "node_modules"
  ]
}
```

Run `$ npm install && npm run build`.

**Expected behavior:**

`dist/foo.d.ts`:

```ts
declare const foo: {
    bar: string;
};
export { foo };
```

`dist/index.d.ts`:

```ts
import { foo } from './foo';
export { foo };
declare const baz: string;
export { baz };
```

**Actual behavior:**

`dist/foo.d.ts`:

```ts
declare const foo: {
    bar: string;
};
export { foo };
```

`dist/index.d.ts`:

```ts
import { foo } from './foo';
export { foo };
export { baz };  // ERROR: Cannot find name `'baz'`.
```

---

Note:

This would fail, too:

```ts
const { bar } = foo;
export { bar };
```

This works:

```ts
const baz = foo.bar;
export { baz };
```

### LINKED ISSUE #18634: When generating definitions re-exporting variables destructured from functions has missing types
**TypeScript Version:** 

2.4.2 / 2.5.2

**Code**

```ts
const getFoo = () => ({
    foo: 'foo'
})

// broken: foo is missing in definitions
const { foo } = getFoo()
export { foo }

// workaround: foo is in definitions
export const { foo } = getFoo()
```

**Expected behavior:**

foo should exist in the generated definitions when destructured from a function and then exported separately.

**Actual behavior:**

foo is missing from the generated definitions when destructured from a function and then exported separately.

### PR DESCRIPTION
**Pull request by weswigham**
`collectLinkedAliases` did some painting of declaration visibility in the declaration emitter. This was bad, as one thing would get painted, then written, then another painted, then written; so if one element referenced another element made visible through an alias, we would issue an error. (And only on the first compile in a program! If emit was called twice, we would succeed!)

Now we do that analysis inside the checker, where it should have been all along.

Fixes #14753
Fixes #17085
Fixes #18634

### CODE REVIEW

**Comment by weswigham** [comment]
Now also included is a fix for #17085, since it could not be fixed until the visibility of exports was fixed.

**Comment by weswigham** [comment]
Also fixes #18634.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -23258,6 +23263,7 @@ namespace ts {
 
         function checkExportSpecifier(node: ExportSpecifier) {
             checkAliasSymbol(node);
+            collectLinkedAliases(node.propertyName || node.name, /*setVisibility*/ true); // Collect linked aliases to set visibility links
```
Comment (mhegazy) [inline src/compiler/checker.ts]: you only need to do this if `--declaration` is set..

File: src/compiler/declarationEmitter.ts [hunk 2]
```diff
@@ -1249,10 +1249,18 @@ namespace ts {
             writeLine();
         }
 
+        function bindingNameContainsVisibleBindingElement(node: BindingName): boolean {
+            return !!node && isBindingPattern(node) && some(node.elements, elem => !isOmittedExpression(elem) && (resolver.isDeclarationVisible(elem) || bindingNameContainsVisibleBindingElement(elem.name)));
```
Comment (mhegazy) [inline src/compiler/declarationEmitter.ts]: `elem => !isOmittedExpression(elem) && isVariableDeclarationVisible(elem)` ?

Comment (weswigham) [reply src/compiler/declarationEmitter.ts]: done

File: src/compiler/checker.ts (hunk 1 above)
Comment (weswigham) [reply src/compiler/checker.ts]: done

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -23258,6 +23263,9 @@ namespace ts {
 
         function checkExportSpecifier(node: ExportSpecifier) {
             checkAliasSymbol(node);
+            if (compilerOptions.declaration) {
+                collectLinkedAliases(node.propertyName || node.name, /*setVisibility*/ true); // Collect linked aliases to set visibility links
```
Comment (sandersn) [inline src/compiler/checker.ts]: both this comment and the other one seem redundant. And if they're not, then things are actually *more* complicated than they appear.

Comment (weswigham) [reply src/compiler/checker.ts]: They didn't feel redundant before I added the `setVisibility` flag. :(

I'll open a PR to remove them.

**Pull request merged**